### PR TITLE
Add authentication for the REST API

### PR DESCRIPTION
**Pull request by RVRX**
This authentication for the Events API works via public key encryption. I **do not expect it to work with anyone's _current_ SSH keys, but any machine with OpenSSL can easily generate keys understood by the API**. There is also currently no way – to my knowledge – to add or update new users in the DB without directly interfacing with it.

### User Authentication Plug
The API now uses a new Authentication Plug, that should carry over to any other route that implements the pipeline `:requires_auth` with only a little modification (probably just adding the the `:raw_body` field via a body_reader parser).

Any request that goes through this pipeline is required to include: 
 - The `?user=<user_name>` query parameter, 
 - `http-auth` header with the base64 encoding of the signature of the body of the  request. Only requests with a valid signature that maps to the public key of user in the DB will be allowed to access the resource on the other end of the pipeline.

Keys must be PEM-format RSA key with SHA256 encryption, such as that generated by OpenSSL. Such a key can be generated with:
```shell
openssl genrsa -out helios.pem && openssl rsa -in helios.pem -pubout -out helios_pub.pem
```
It might be possible to generate such keys with ssh-keygen, but I did not stumble upon a valid output key.

### The post-commit Hook
Might have some optimizations to make in this script, but when placed in `.git/hooks` it will fire after every commit you make in the given repository.

There are two variables near the top of this script that should be modified to point to your private key and a Helios instance.

**Note**: when in a Helios development environment, with the hook pointing to your localhost you might receive some noisy console output upon committing when the repository state is such that errors are produced by Phoenix, as Phoenix will respond with an apparently unsilenceable HTML response stack-trace.

### CODE REVIEW

**Review by cathyzoller — COMMENTED** [review]
Nice work... needs a rebase before approval, and I have just a couple of comments.

File: server-phoenix/lib/helios_web/plugs/user_authentication.ex
Comment (cbarber) [inline server-phoenix/lib/helios_web/plugs/user_authentication.ex:53]: Could we put required fields in the header? Would this simplify the git hook?
```
Authorization: Signature user="user",signature="base64-payload"
```

File: server-phoenix/test/helios_web/controllers/events_api/events_api_controller_test.exs
Comment (cbarber) [inline server-phoenix/test/helios_web/controllers/events_api/events_api_controller_test.exs:20]: Could you add a comment at the top of the file for how to re-create these payloads from the CLI (assuming that's how you did it)?

File: server-phoenix/test/helios_web/controllers/events_api/pub.pem
Comment (cbarber) [inline server-phoenix/test/helios_web/controllers/events_api/pub.pem:9]: Lets add the private key as well in case we need to recreate any of the above payloads (assuming these are both created just for this test). If this is your actual public key, lets switch it to one we create just for testing.

File: server-phoenix/test/helios_web/controllers/events_api/events_api_controller_test.exs
Comment (cbarber) [inline server-phoenix/test/helios_web/controllers/events_api/events_api_controller_test.exs:101]: :thinking: I feel like these were useful tests. If you add the throw away private key, can you sign them programmatically to continue to use them?

File: post-commit
Comment (cbarber) [inline post-commit:58]: IIRC, openssl will sign STDIN in if you omit the file name. Will let you avoid the temp file.

**Review by chewnoill — APPROVED** [review]

File: server-phoenix/lib/helios_web/plugs/user_authentication.ex
Comment (RVRX) [reply server-phoenix/lib/helios_web/plugs/user_authentication.ex:53]: So the username is already in the JSON body, so it looks like I can effective do the same thing as JWT (and doing what barber said) by refactoring the auth to get the username from the POST body insetad of the query param.

So now...
 - user sends POST,
 - Auth looks at POST, gets the user
 - gets that user's key
 - verify the body against the signature
Which effectively is the same as JWT except we've got an unencoded body instead of an encoded (b64) header.

File: server-phoenix/test/helios_web/controllers/events_api/events_api_controller_test.exs
Comment (RVRX) [reply server-phoenix/test/helios_web/controllers/events_api/events_api_controller_test.exs:101]: I kinda felt like they were redundant, but was maybe biased by the chore of not wanting to sign and encrypt anymore payloads. I'll add them back

File: server-phoenix/lib/helios_web/plugs/user_authentication.ex
Comment (cbarber) [reply server-phoenix/lib/helios_web/plugs/user_authentication.ex:53]: Sounds fine. If we want to explore standards for either the auth header (Signature has an rfc) or JWTs that can be a future iteration on this so we don't scope creep.

File: server-phoenix/test/helios_web/controllers/events_api/pub.pem
Comment (RVRX) [reply server-phoenix/test/helios_web/controllers/events_api/pub.pem:9]: added

File: post-commit
Comment (RVRX) [reply post-commit]: yup, fixed

**Pull request merged**
